Constructions/eps_additon.ml

```diff
@@ -0,0 +1,2136 @@
+  eval (add_ebin (QuoConst "One" (TyBase "nat")) x) to (nat) = 
+  add_unary One (eval x to (nat))`,
+  MATCH_MP_TAC(lth) THEN 
+  CONJ_TAC THENL
+  [REWRITE_TAC[proper_nat_construct]
+  ;CONJ_TAC THENL
+  [REPEAT GEN_TAC THEN 
+  REWRITE_TAC[thm3_x_quoconst] THEN 
+  REWRITE_TAC[proper_nat_construct;IMP_CONJ] THEN 
+  REPEAT DISCH_TAC THEN 
+  DISJ_CASES_TAC(ASSUME `a0 = "Zero" \/ a0 = "One"`) THEN 
+  ASM_REWRITE_TAC[add_ebin] THEN 
+  CONSTRUCTION_TO_QUOTE_TAC THEN 
+  LAW_OF_DISQUO_TAC THEN 
+  REWRITE_TAC[add_unary;thenZero;one_def]
+  ;CONJ_TAC THENL
+  [REPEAT GEN_TAC THEN 
+  REWRITE_TAC[sub_a1_thm;sub_app_thm;IMP_CONJ] THEN 
+  DISCH_TAC THEN 
+  DISCH_TAC THEN 
+  REWRITE_TAC[proper_nat_construct;IMP_CONJ] THEN 
+  REPEAT DISCH_TAC THEN 
+  DISJ_CASES_TAC(ASSUME 
+    `(a0 = QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat")) \/
+    a0 = QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat")))`) THEN 
+  ASM_REWRITE_TAC[add_ebin] THENL
+  [MP_TAC(APP_DISQUO 
+    `QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `a1:epsilon`) THEN 
+  MP_TAC(APP_DISQUO 
+    `QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `a1:epsilon`) THEN 
+  PROPER_TYPE_TAC(`a1:epsilon`) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  IS_EXPR_TYPE_TAC THEN
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  CONSTRUCTION_TO_QUOTE_TAC THEN 
+  LAW_OF_DISQUO_TAC THEN 
+  REWRITE_TAC[one_def;take_s;id_of_plus;add_one_even]
+  ;MP_TAC(APP_DISQUO 
+    `QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `a1:epsilon`) THEN 
+  MP_TAC(APP_DISQUO 
+    `QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))` 
+    `(add_ebin (QuoConst "One" (TyBase "nat")) a1):epsilon`) THEN 
+  PROPER_TYPE_TAC(`a1:epsilon`) THEN
+  SUBGOAL_THEN 
+    `proper_nat_construct (QuoConst "One" (TyBase "nat"))` ASSUME_TAC THEN 
+  REWRITE_TAC[proper_nat_construct] THEN 
+  MP_TAC(SPECL [`(QuoConst "One" (TyBase "nat"))`;`a1:epsilon`] thm1) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  IS_EXPR_TYPE_TAC THEN
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[] THEN 
+  CONSTRUCTION_TO_QUOTE_TAC THEN 
+  LAW_OF_DISQUO_TAC THEN 
+  REWRITE_TAC[one_def;take_s;id_of_plus;add_one_even] THEN 
+  MP_TAC(ASSUME `proper_nat_construct a1
+  ==> (eval (add_ebin (QuoConst "One" (TyBase "nat")) a1) to (nat)) =
+  add_unary One (eval (a1) to (nat))`) THEN 
+  REWRITE_TAC[ASSUME `proper_nat_construct a1`] THEN 
+  DISCH_TAC THEN 
+  QUOTE_TO_CONSTRUCTION_TAC THEN
+  ASM_REWRITE_TAC[
+    carry_one;one_def;take_s;id_of_plus;add_one_even;take_out_S;odd_to_even]
+  ]
+  ;REWRITE_TAC[proper_nat_construct]
+  ]
+  ]
+  ]
+  );;
+
+
+
+(*                                                     *)
+(* Meaning formula for addition                        *)
+(*                                                     *)
+
+let add_meaning = 
+  let tm = 
+    `(!x:epsilon y:epsilon. 
```

Comment:
This term embodies the shape idea I was mentioning yesterday.  Let T be a meta-variable for a type, F a binary syntax transformer and G a binary function (that are supposed to correspond), then your term has shape
```
`(!x:epsilon y:epsilon.
   (((proper_X_construct x) /\ (proper_X_construct y))
   ==> (eval (F x y) to (X) =
   G (eval x to (X)) (eval y to (X))))))`
```
This also points out that putting X in the name of `proper_construct` may be limiting.  @wmfarmer should ponder on this too.